File: typescript/srp.ts
// SRP (Single Responsibility Principle)
// Separate the code into simple, well-defined, well-intentioned tasks and give clear names.

// Before
class ReportGenerator {
  generateReport(data: string[]): void {
    // Process the data
    const processedData = data.map(item => item.toUpperCase());

    // Convert data to JSON
    const jsonData = JSON.stringify(processedData);

    // Save the report to a file (simulated)
    console.log(`Saving to file: ${jsonData}`);

    // Send the report via email (simulated)
    console.log(`Email sent with data: ${jsonData}`);
  }
}

const data = ["a", "b", "c", "d"];
const report = new ReportGenerator();
report.generateReport(data);


// After
class ReportGenerator {
  generateReport(data: string[]): void {
    const processedData = this._processData(data);
    const jsonData = this._convertToJson(processedData);
    this._saveReport(jsonData);
    this._sendReport(jsonData);
  }

  private _processData(data: string[]): string[] {
    return data.map(item => item.toUpperCase());
  }

  private _convertToJson(data: string[]): string {
    return JSON.stringify(data);
  }

  private _saveReport(jsonData: string): void {
    console.log(`Saving to file: ${jsonData}`);
  }

  private _sendReport(jsonData: string): void {
    console.log(`Email sent with data: ${jsonData}`);
  }
}

const data = ["a", "b", "c", "d"];
const report = new ReportGenerator();
report.generateReport(data);

File: rust/srp.rs
// SRP (Single Responsibility Principle)
// Separate the code into simple, well-defined, well-intentioned tasks and give clear names.

// Before
struct ReportGeneratorBefore;

impl ReportGeneratorBefore {
    fn generate_report(&self, data: Vec<String>) {
        // Process the data
        let processed_data: Vec<String> = data.iter().map(|item| item.to_uppercase()).collect();

        // Convert data to JSON (simplified)
        let json_data = format!("[{}]", processed_data.iter()
            .map(|item| format!("\"{}\"", item))
            .collect::<Vec<_>>()
            .join(", "));

        // Save the report to a file (simulated)
        println!("Saving to file: {}", json_data);

        // Send the report via email (simulated)
        println!("Email sent with data: {}", json_data);
    }
}

// After
struct ReportGenerator;

impl ReportGenerator {
    fn generate_report(&self, data: Vec<String>) {
        let processed_data = self.process_data(data);
        let json_data = self.convert_to_json(processed_data);
        self.save_report(&json_data);
        self.send_report(&json_data);
    }

    fn process_data(&self, data: Vec<String>) -> Vec<String> {
        data.iter().map(|item| item.to_uppercase()).collect()
    }

    fn convert_to_json(&self, data: Vec<String>) -> String {
        format!("[{}]", data.iter()
            .map(|item| format!("\"{}\"", item))
            .collect::<Vec<_>>()
            .join(", "))
    }

    fn save_report(&self, json_data: &str) {
        println!("Saving to file: {}", json_data);
    }

    fn send_report(&self, json_data: &str) {
        println!("Email sent with data: {}", json_data);
    }
}

fn main() {
    let data = vec!["a".to_string(), "b".to_string(), "c".to_string(), "d".to_string()];
    
    let report_before = ReportGeneratorBefore;
    report_before.generate_report(data.clone());

    println!();

    let report = ReportGenerator;
    report.generate_report(data);
}
